backends/concrete-cpu/implementation/src/c_api/fheint.rs

```diff
@@ -0,0 +1,207 @@
+use super::utils::nounwind;
+use core::slice;
+use std::io::Cursor;
+use tfhe::core_crypto::prelude::*;
+use tfhe::integer::ciphertext::Expandable;
+use tfhe::integer::IntegerCiphertext;
+use tfhe::shortint::parameters::{Degree, NoiseLevel};
+use tfhe::shortint::{CarryModulus, Ciphertext, MessageModulus};
+use tfhe::FheUint8;
+
+#[repr(C)]
+pub struct TfhersFheIntDescription {
+    width: usize,
+    is_signed: bool,
+    lwe_size: usize,
+    n_cts: usize,
+    degree: usize,
+    noise_level: usize,
+    message_modulus: usize,
+    carry_modulus: usize,
+    ks_first: bool,
+}
+
+impl TfhersFheIntDescription {
+    fn zero() -> TfhersFheIntDescription {
+        TfhersFheIntDescription {
+            width: 0,
+            is_signed: false,
+            lwe_size: 0,
+            n_cts: 0,
+            degree: 0,
+            noise_level: 0,
+            message_modulus: 0,
+            carry_modulus: 0,
+            ks_first: false,
+        }
+    }
+
+    /// Create a `Ciphertext` from the lwe and `self` metadata
+    fn ct_from_lwe(&self, lwe: LweCiphertext<Vec<u64>>) -> Ciphertext {
+        // all this if/else due to the fact that we can't construct a NoiseLevel (not public)
+        let noise_level = if self.noise_level == NoiseLevel::ZERO.get() {
+            NoiseLevel::ZERO
+        } else if self.noise_level == NoiseLevel::NOMINAL.get() {
+            NoiseLevel::NOMINAL
+        } else if self.noise_level == NoiseLevel::MAX.get() {
+            NoiseLevel::MAX
+        } else if self.noise_level == NoiseLevel::UNKNOWN.get() {
+            NoiseLevel::UNKNOWN
+        } else {
+            panic!("invalid noise level value");
+        };
+
+        Ciphertext::new(
+            lwe,
+            Degree::new(self.degree),
+            noise_level,
+            MessageModulus(self.message_modulus),
+            CarryModulus(self.carry_modulus),
+            if self.ks_first {
+                PBSOrder::KeyswitchBootstrap
+            } else {
+                PBSOrder::BootstrapKeyswitch
+            },
+        )
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn concrete_cpu_tfhers_unknown_noise_level() -> usize {
+    NoiseLevel::UNKNOWN.get()
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn concrete_cpu_tfhers_uint8_description(
+    serialized_data_ptr: *const u8,
+    serialized_data_len: usize,
+) -> TfhersFheIntDescription {
+    nounwind(|| {
+        // deserialize fheuint8
+        let mut serialized_data = Cursor::new(slice::from_raw_parts(
+            serialized_data_ptr,
+            serialized_data_len,
+        ));
+        let fheuint: FheUint8 = match bincode::deserialize_from(&mut serialized_data) {
```

Comment:
to be future proof and benefit from the data compatibility you would want to use the versionize primitives before serializing, then you are free to use whichever serde-compatible serialization backend

Reply:
see e.g. here https://github.com/zama-ai/tfhe-rs/tree/main/utils/tfhe-versionable Nicolas is the one who worked on the design of this if support is needed

Reply:
I don't think it's TFHErs vs Concrete versions. It's the user of TFHErs which will be doing serialization, so he should make sure he uses the same version as in Concrete. I don't think TFHErs has something to do in here. For example, you can see the tool I'm using for testing (tfhers-utils). It's using bincode for serialization and doesn't rely on TFHErs for that. Unless serialization of TFHErs struct would rely on something implemeneted in TFHErs that I'm not aware of.

Reply:
@IceTDrinker I was not worried by the versionning of TFHE-rs structures, but more in the compatibility of the serialization format implemented by bincode. 

@youben11 you are right, but Concrete is delivered as a binary (with a fixed version for bincode), but we don't know what users will do in their rust code, nor if they will be aware that the version of bincode is important.

Reply:
bincode format should be stable and if they break it I do hope they release a major version 😅

Reply:
@aPere3 yes but in that case, I expect Concrete users to follow the framework instead of using a different serialization format.
@IceTDrinker we should also move to the new versioned serialization soon

Reply:
@IceTDrinker Not sure ! I did not go into many details but it seems they changed their spec since last major release (https://github.com/bincode-org/bincode/commits/trunk/docs/spec.md). 

@youben11 what do you mean by `follow the framework` ? The basic user (me) won't look at the repo tree at the commit of the concrete binary to extract the `bincode` version. He will just do `Cargo.toml bincode="2"` and call it a day 🤣 

I any case, I am not overly worried about that 😉 Though it may be interesting to think about adding bincode version in the serialization if they are not reliable.

Reply:
I mean we should document that. Otherwise, we should provide ser/deser functions maybe? I'm no expert in Rust serialization, but I would always think how to serialize something before sending it to another package for deserialization. Or is bincode what everyone uses? Think of when we will be using versionize: we will have to document how someone is expected to serialize a ciphertext before using it in Concrete.

Reply:
Actually ciphertexts will come from tfhe-rs users (blockchain as example). So we should use the same serialization framework than casual tfhe-rs user will use, if I well understand the versionable framework is something generic, but is that will be used for tfhe-rs integers @IceTDrinker ? From my understanding you should implement versioning on tfhe-rs side for thf-rs integers and we should use that.

Reply:
we already have versioning for ciphertexts, Nicolas will be back on thursday so if Ayoub needs a quick tour it can likely be done then

Reply:
Ok great :+1: